File: AlgoWorks/src/com/solid/algolearning/javacode/sorting/SelectionSort.java
package com.solid.algolearning.javacode.sorting;

import java.util.Arrays;

public class SelectionSort {

//    Selection sort implementation
    public static void main(String[] args) {
        int[] numbers = {64, 25, 12, 22, 11};

        selectSort(numbers, numbers.length);

    }

    public static void selectSort(int[] numbers, int n){
        for(int lastUnsortedIndex = n - 1; lastUnsortedIndex > 0; lastUnsortedIndex--){
           int largest = 0;
            for (int i = 1; i <= lastUnsortedIndex; i++){
                if (numbers[i] > numbers[largest]){
                    largest = i;
                }
            }

            swap(numbers, largest, lastUnsortedIndex);
        }

        printOut(numbers);
    }

    public static void swap(int[] array, int i, int j){
        if (i == j){
            return;
        }

        int temp = array[i];
        array[i] = array[j];
        array[j] = temp;
    }

    public static void printOut(int[] array){
        System.out.println(Arrays.toString(array));
    }
}

File: AlgoWorks/src/com/solid/algolearning/kotlincode/algorithms/sorting/SelectionSort.kt
package com.solid.algolearning.kotlincode.algorithms.sorting

import com.solid.algolearning.javacode.algorithms.sorting.SelectionSort
import java.util.*

// Selection sort algorithm works by scanning for the smallest item in the list and replacing them with the item in the first index
fun main() {

    val numbers = intArrayOf(64, 25, 12, 22, 11)

    selectSort(numbers, numbers.size)
}

fun selectSort(numbers: IntArray, n: Int) {
    for (lastUnsortedIndex in n - 1 downTo 1) {
        var largest = 0
        for (i in 1..lastUnsortedIndex) {
            if (numbers[i] > numbers[largest]) {
                largest = i
            }
        }
        swap(numbers, largest, lastUnsortedIndex)
    }
    printOut(numbers)
}

fun swap(array: IntArray, i: Int, j: Int) {
    if (i == j) {
        return
    }
    val temp = array[i]
    array[i] = array[j]
    array[j] = temp
}

fun printOut(array: IntArray?) {
    println(Arrays.toString(array))
}